File: ldk-nbt/src/main/java/com/lgou2w/ldk/nbt/NBTTagList.java
/*
 * Copyright (C) 2016-2021 The lgou2w <lgou2w@hotmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.lgou2w.ldk.nbt;

import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;

import java.io.DataInput;
import java.io.DataOutput;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.Iterator;
import java.util.List;
import java.util.ListIterator;
import java.util.Objects;

public class NBTTagList extends NBTBase<List<NBTBase<?>>> implements List<NBTBase<?>> {
  private @NotNull NBTType elementType = NBTType.END;

  @Contract("null, _ -> fail; _, null -> fail")
  public NBTTagList(String name, List<NBTBase<?>> value) {
    super(name, new ArrayList<>(value));
    for (NBTBase<?> el : value) checkElement(el);
  }

  @Contract("null -> fail")
  public NBTTagList(String name) {
    this(name, new ArrayList<>());
  }

  @Contract("null -> fail")
  public NBTTagList(List<NBTBase<?>> value) {
    this("", value);
  }

  public NBTTagList() {
    this("", new ArrayList<>());
  }

  private void checkElement(@NotNull NBTBase<?> el) {
    if (elementType == NBTType.END) {
      elementType = el.getType();
    } else if (el.getType() != elementType) {
      throw new IllegalArgumentException("The type of element value that does not match, should be: " + elementType);
    }
  }

  @NotNull
  public NBTType getElementType() {
    return elementType;
  }

  @Override
  @NotNull
  public NBTType getType() {
    return NBTType.LIST;
  }

  @Override
  public void setValue(List<NBTBase<?>> value) {
    if (value == null) throw new NullPointerException("value");
    for (NBTBase<?> el : value) checkElement(el);
    super.setValue(new ArrayList<>(value));
  }

  @Override
  public void read(@NotNull DataInput input) throws IOException {
    NBTType elementType = NBTType.fromId(input.readUnsignedByte());
    if (elementType == null) elementType = NBTType.END;
    int length = input.readInt();
    this.elementType = elementType;
    value = new ArrayList<>();
    for (int i = 0; i < length; i++) {
      NBTBase<?> element = NBTType.create(elementType);
      element.read(input);
      value.add(element); // skip check
    }
  }

  @Override
  public void write(@NotNull DataOutput output) throws IOException {
    output.writeByte(isEmpty() ? 0 : elementType.getId());
    output.writeInt(size());
    for (NBTBase<?> el : value) el.write(output);
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;
    NBTTagList list = (NBTTagList) o;
    if (elementType != list.elementType) return false;
    return super.equals(o);
  }

  @Override
  public int hashCode() {
    return Objects.hash(super.hashCode(), elementType);
  }

  @Override
  public String toString() {
    return "NBTTagList{" +
      "value=" + value +
      ", elementType=" + elementType +
      '}';
  }

  @Override
  protected void toMojangsonBuilder(@NotNull StringBuilder builder, boolean color) {
    builder.append('[');
    int len = value.size(), i = 0;
    for (NBTBase<?> element : value) {
      if (i >= 1 && i < len) builder.append(!color ? "," : ", ");
      element.toMojangsonBuilder(builder, color);
      i++;
    }
    builder.append(']');
  }

  @Override
  @NotNull
  public NBTTagList clone() {
    List<NBTBase<?>> newValue = new ArrayList<>(value.size());
    for (NBTBase<?> el : value) newValue.add(el.clone());
    return new NBTTagList(name, newValue);
  }

  /// Extended

  @Contract("null -> fail")
  public boolean addByte(byte... values) {
    if (values == null) throw new NullPointerException("values");
    List<NBTTagByte> c = new ArrayList<>(values.length);
    for (byte element : values) c.add(new NBTTagByte(element));
    return addAll(c);
  }

  @Contract("null -> fail")
  public boolean addByte(int... values) {
    if (values == null) throw new NullPointerException("values");
    List<NBTTagByte> c = new ArrayList<>(values.length);
    for (int element : values) c.add(new NBTTagByte((byte) element));
    return addAll(c);
  }

  @Contract("null -> fail")
  public boolean addShort(short... values) {
    if (values == null) throw new NullPointerException("values");
    List<NBTTagShort> c = new ArrayList<>(values.length);
    for (short element : values) c.add(new NBTTagShort(element));
    return addAll(c);
  }

  @Contract("null -> fail")
  public boolean addShort(int... values) {
    if (values == null) throw new NullPointerException("values");
    List<NBTTagShort> c = new ArrayList<>(values.length);
    for (int element : values) c.add(new NBTTagShort((short) element));
    return addAll(c);
  }

  @Contract("null -> fail")
  public boolean addInt(int... values) {
    if (values == null) throw new NullPointerException("values");
    List<NBTTagInt> c = new ArrayList<>(values.length);
    for (int element : values) c.add(new NBTTagInt(element));
    return addAll(c);
  }

  @Contract("null -> fail")
  public boolean addLong(long... values) {
    if (values == null) throw new NullPointerException("values");
    List<NBTTagLong> c = new ArrayList<>(values.length);
    for (long element : values) c.add(new NBTTagLong(element));
    return addAll(c);
  }

  @Contract("null -> fail")
  public boolean addFloat(float... values) {
    if (values == null) throw new NullPointerException("values");
    List<NBTTagFloat> c = new ArrayList<>(values.length);
    for (float element : values) c.add(new NBTTagFloat(element));
    return addAll(c);
  }

  @Contract("null -> fail")
  public boolean addDouble(double... values) {
    if (values == null) throw new NullPointerException("values");
    List<NBTTagDouble> c = new ArrayList<>(values.length);
    for (double element : values) c.add(new NBTTagDouble(element));
    return addAll(c);
  }

  @Contract("null -> fail")
  public boolean addByteArray(byte[]... values) {
    if (values == null) throw new NullPointerException("values");
    List<NBTTagByteArray> c = new ArrayList<>(values.length);
    for (byte[] element : values) c.add(new NBTTagByteArray(element));
    return addAll(c);
  }

  @Contract("null -> fail")
  public boolean addString(String... values) {
    if (values == null) throw new NullPointerException("values");
    List<NBTTagString> c = new ArrayList<>(values.length);
    for (String element : values) c.add(new NBTTagString("", element));
    return addAll(c);
  }

  @Contract("null -> fail")
  public boolean addList(NBTTagList... values) {
    return addAll(Arrays.asList(values));
  }

  @Contract("null -> fail")
  public boolean addCompound(NBTTagCompound... values) {
    return addAll(Arrays.asList(values));
  }

  @Contract("null -> fail")
  public boolean addIntArray(int[]... values) {
    if (values == null) throw new NullPointerException("values");
    List<NBTTagIntArray> c = new ArrayList<>(values.length);
    for (int[] element : values) c.add(new NBTTagIntArray(element));
    return addAll(c);
  }

  @Contract("null -> fail")
  public boolean addLongArray(long[]... values) {
    if (values == null) throw new NullPointerException("values");
    List<NBTTagLongArray> c = new ArrayList<>(values.length);
    for (long[] element : values) c.add(new NBTTagLongArray(element));
    return addAll(c);
  }

  @Contract("null -> fail")
  public boolean addBoolean(boolean... values) {
    if (values == null) throw new NullPointerException("values");
    List<NBTTagByte> c = new ArrayList<>(values.length);
    for (boolean element : values) c.add(new NBTTagByte((byte) (element ? 1 : 0)));
    return addAll(c);
  }

  /// List

  @NotNull
  @Override
  public Iterator<NBTBase<?>> iterator() {
    return value.iterator();
  }

  @Override
  public int size() {
    return value.size();
  }

  @Override
  public boolean contains(Object o) {
    return value.contains(o);
  }

  @Override
  public boolean containsAll(@NotNull Collection<?> c) {
    return value.containsAll(c);
  }

  @Override
  public NBTBase<?> get(int index) {
    return value.get(index);
  }

  @Override
  public int indexOf(Object o) {
    return value.indexOf(o);
  }

  @Override
  public boolean isEmpty() {
    return value.isEmpty();
  }

  @Override
  public int lastIndexOf(Object o) {
    return value.lastIndexOf(o);
  }

  @Override
  public boolean add(NBTBase<?> element) {
    checkElement(element);
    return value.add(element);
  }

  @Override
  public void add(int index, NBTBase<?> element) {
    checkElement(element);
    value.add(index, element);
  }

  @Override
  public boolean addAll(int index, @NotNull Collection<? extends NBTBase<?>> c) {
    for (NBTBase<?> el : c) checkElement(el);
    return value.addAll(index, c);
  }

  @Override
  public boolean addAll(@NotNull Collection<? extends NBTBase<?>> c) {
    for (NBTBase<?> el : c) checkElement(el);
    return value.addAll(c);
  }

  @Override
  public void clear() {
    value.clear();
  }

  @NotNull
  @Override
  public ListIterator<NBTBase<?>> listIterator() {
    return value.listIterator();
  }

  @NotNull
  @Override
  public ListIterator<NBTBase<?>> listIterator(int index) {
    return value.listIterator(index);
  }

  @Override
  public NBTBase<?> remove(int index) {
    return value.remove(index);
  }

  @Override
  public boolean remove(Object o) {
    if (o instanceof NBTBase) checkElement((NBTBase<?>) o);
    return value.remove(o);
  }

  @Override
  public boolean removeAll(@NotNull Collection<?> c) {
    for (Object el : c) {
      if (el instanceof NBTBase) {
        checkElement((NBTBase<?>) el);
      }
    }
    return value.removeAll(c);
  }

  @Override
  public boolean retainAll(@NotNull Collection<?> c) {
    for (Object el : c) {
      if (el instanceof NBTBase) {
        checkElement((NBTBase<?>) el);
      }
    }
    return value.retainAll(c);
  }

  @Override
  public NBTBase<?> set(int index, NBTBase<?> element) {
    checkElement(element);
    return value.set(index, element);
  }

  @NotNull
  @Override
  public List<NBTBase<?>> subList(int fromIndex, int toIndex) {
    return value.subList(fromIndex, toIndex);
  }

  @NotNull
  @Override
  public Object[] toArray() {
    return value.toArray();
  }

  @NotNull
  @Override
  public <T> T[] toArray(@NotNull T[] a) {
    return value.toArray(a);
  }
}

File: ldk-nbt/src/main/kotlin/com/lgou2w/ldk/nbt/NBTTagList.kt
/*
 * Copyright (C) 2018 The lgou2w <lgou2w@hotmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.lgou2w.ldk.nbt

import java.io.DataInput
import java.io.DataOutput
import java.util.*

/**
 * ## NBTTagList (集合 NBT 标签)
 *
 * @see [List]
 * @author lgou2w
 */
class NBTTagList : NBTBase<MutableList<NBTBase<*>>>, MutableList<NBTBase<*>> {

    /**
     * * The NBT tag type within this collection.
     * * 此集合内的 NBT 标签类型.
     *
     * @see [NBTType]
     */
    var elementType = NBTType.TAG_END
        private set

    @JvmOverloads
    constructor(name: String = "", value: MutableList<NBTBase<*>> = LinkedList()) : super(name, value) {
        for (el in value)
            check(el)
    }

    constructor(value: MutableList<NBTBase<*>> = LinkedList()) : super("", value) {
        for (el in value)
            check(el)
    }

    override var value: MutableList<NBTBase<*>>
        get() = super.value0
        set(value) {
            for (el in value)
                check(el)
            super.value0 = ArrayList(value)
        }

    private fun check(element: NBTBase<*>) {
        if (elementType == NBTType.TAG_END)
            elementType = element.type
        else if (elementType != element.type)
            throw IllegalArgumentException("The type of list element value that does not match, should be: $elementType.")
    }

    override val type: NBTType
        get() = NBTType.TAG_LIST

    override fun read(input: DataInput) {
        val elementType = NBTType.fromId(input.readUnsignedByte()) ?: NBTType.TAG_END
        val length = input.readInt()
        this.elementType = elementType
        super.value0 = ArrayList()
        (0 until length).forEach {
            val tag = NBTType.createTag(elementType)
            tag.read(input)
            add(tag)
        }
    }

    override fun write(output: DataOutput) {
        output.writeByte(if (isEmpty()) 0 else elementType.id)
        output.writeInt(size)
        for (el in value0)
            el.write(output)
    }

    override fun clone(): NBTTagList {
        val value = value0.map { it.clone() }.toMutableList()   // clone
        return NBTTagList(name, value)
    }

    override fun hashCode(): Int {
        return super.hashCode() xor elementType.hashCode()
    }

    override fun equals(other: Any?): Boolean {
        if (super.equals(other) && (other as NBTTagList).elementType == elementType)
            return other.value0 == value0
        return false
    }

    override fun toString(): String {
        return "NBTTagList(${toJson()})"
    }

    override fun toJson(): String {
        return buildString {
            append("[")
            for ((index, el) in value0.withIndex()) {
                if (index > 0)
                    append(",")
                append(el.toJson())
            }
            append("]")
        }
    }

    override fun toMojangson(): String {
        return buildString {
            append("[")
            for ((index, el) in value0.withIndex()) {
                if (index > 0)
                    append(",")
                append(el.toMojangson())
            }
            append("]")
        }
    }

    override fun toMojangsonWithColor(): String {
        return buildString {
            append("[")
            for ((index, el) in value0.withIndex()) {
                if (index > 0)
                    append(", ") // one space blank
                append(el.toMojangsonWithColor())
            }
            append("]")
        }
    }

    ///
    //

    @Throws(ClassCastException::class)
    fun <T> asElements() : List<T> {
        @Suppress("UNCHECKED_CAST")
        return if (elementType.isWrapper())
            value0.map { it as T }
        else
            value0.map { it.value as T }
    }

    /**
     * * Add the given byte [value] to this collection tag.
     * * 将给定的字节值 [value] 添加到此集合标签内.
     *
     * @throws IllegalArgumentException If the collection element type and parameter type do not match.
     * @throws IllegalArgumentException 如果集合元素类型和参数类型不匹配.
     */
    fun addByte(vararg value: Byte): Boolean {
        return addAll(value.map { NBTTagByte(it) })
    }

    /**
     * * Add the given byte [value] to this collection tag.
     * * 将给定的字节值 [value] 添加到此集合标签内.
     *
     * @throws IllegalArgumentException If the collection element type and parameter type do not match.
     * @throws IllegalArgumentException 如果集合元素类型和参数类型不匹配.
     */
    fun addByte(vararg value: Int): Boolean {
        return addAll(value.map { NBTTagByte(it.toByte()) })
    }

    /**
     * * Add the given short [value] to this collection tag.
     * * 将给定的短整数值 [value] 添加到此集合标签内.
     *
     * @throws IllegalArgumentException If the collection element type and parameter type do not match.
     * @throws IllegalArgumentException 如果集合元素类型和参数类型不匹配.
     */
    fun addShort(vararg value: Short): Boolean {
        return addAll(value.map { NBTTagShort(it) })
    }

    /**
     * * Add the given short [value] to this collection tag.
     * * 将给定的短整数值 [value] 添加到此集合标签内.
     *
     * @throws IllegalArgumentException If the collection element type and parameter type do not match.
     * @throws IllegalArgumentException 如果集合元素类型和参数类型不匹配.
     */
    fun addShort(vararg value: Int): Boolean {
        return addAll(value.map { NBTTagShort(it.toShort()) })
    }

    /**
     * * Add the given int [value] to this collection tag.
     * * 将给定的整数值 [value] 添加到此集合标签内.
     *
     * @throws IllegalArgumentException If the collection element type and parameter type do not match.
     * @throws IllegalArgumentException 如果集合元素类型和参数类型不匹配.
     */
    fun addInt(vararg value: Int): Boolean {
        return addAll(value.map { NBTTagInt(it) })
    }

    /**
     * * Add the given long [value] to this collection tag.
     * * 将给定的长整数值 [value] 添加到此集合标签内.
     *
     * @throws IllegalArgumentException If the collection element type and parameter type do not match.
     * @throws IllegalArgumentException 如果集合元素类型和参数类型不匹配.
     */
    fun addLong(vararg value: Long): Boolean {
        return addAll(value.map { NBTTagLong(it) })
    }

    /**
     * * Add the given float [value] to this collection tag.
     * * 将给定的单精度浮点值 [value] 添加到此集合标签内.
     *
     * @throws IllegalArgumentException If the collection element type and parameter type do not match.
     * @throws IllegalArgumentException 如果集合元素类型和参数类型不匹配.
     */
    fun addFloat(vararg value: Float): Boolean {
        return addAll(value.map { NBTTagFloat(it) })
    }

    /**
     * * Add the given double [value] to this collection tag.
     * * 将给定的双精度浮点值 [value] 添加到此集合标签内.
     *
     * @throws IllegalArgumentException If the collection element type and parameter type do not match.
     * @throws IllegalArgumentException 如果集合元素类型和参数类型不匹配.
     */
    fun addDouble(vararg value: Double): Boolean {
        return addAll(value.map { NBTTagDouble(it) })
    }

    /**
     * * Add the given byte array [value] to this collection tag.
     * * 将给定的字节数组值 [value] 添加到此集合标签内.
     *
     * @throws IllegalArgumentException If the collection element type and parameter type do not match.
     * @throws IllegalArgumentException 如果集合元素类型和参数类型不匹配.
     */
    fun addByteArray(vararg value: ByteArray): Boolean {
        return addAll(value.map { NBTTagByteArray(it) })
    }

    /**
     * * Add the given string [value] to this collection tag.
     * * 将给定的字符串值 [value] 添加到此集合标签内.
     *
     * @throws IllegalArgumentException If the collection element type and parameter type do not match.
     * @throws IllegalArgumentException 如果集合元素类型和参数类型不匹配.
     */
    fun addString(vararg value: String): Boolean {
        return addAll(value.map { NBTTagString(it) })
    }

    /**
     * * Add the given list tag [value] to this collection tag.
     * * 将给定的集合标签值 [value] 添加到此集合标签内.
     *
     * @throws IllegalArgumentException If the collection element type and parameter type do not match.
     * @throws IllegalArgumentException 如果集合元素类型和参数类型不匹配.
     */
    fun addList(vararg value: NBTTagList): Boolean {
        return addAll(value)
    }

    /**
     * * Add the given compound tag [value] to this collection tag.
     * * 将给定的复合标签值 [value] 添加到此集合标签内.
     *
     * @throws IllegalArgumentException If the collection element type and parameter type do not match.
     * @throws IllegalArgumentException 如果集合元素类型和参数类型不匹配.
     */
    fun addCompound(vararg value: NBTTagCompound): Boolean {
        return addAll(value)
    }

    /**
     * * Add the given int array [value] to this collection tag.
     * * 将给定的整数数组值 [value] 添加到此集合标签内.
     *
     * @throws IllegalArgumentException If the collection element type and parameter type do not match.
     * @throws IllegalArgumentException 如果集合元素类型和参数类型不匹配.
     */
    fun addIntArray(vararg value: IntArray): Boolean {
        return addAll(value.map { NBTTagIntArray(it) })
    }

    /**
     * * Add the given boolean [value] to this collection tag.
     * * 将给定的布尔值 [value] 添加到此集合标签内.
     *
     * @throws IllegalArgumentException If the collection element type and parameter type do not match.
     * @throws IllegalArgumentException 如果集合元素类型和参数类型不匹配.
     */
    fun addBoolean(vararg value: Boolean): Boolean {
        return addAll(value.map { NBTTagByte(if (it) 1.toByte() else 0.toByte()) })
    }

    //
    ///

    override fun iterator(): MutableIterator<NBTBase<*>> {
        return value0.iterator()
    }

    override val size: Int
        get() = value0.size

    override fun contains(element: NBTBase<*>): Boolean {
        return value0.contains(element)
    }

    override fun containsAll(elements: Collection<NBTBase<*>>): Boolean {
        return value0.containsAll(elements)
    }

    override fun get(index: Int): NBTBase<*> {
        return value0[index]
    }

    override fun indexOf(element: NBTBase<*>): Int {
        return value0.indexOf(element)
    }

    override fun isEmpty(): Boolean {
        return value0.isEmpty()
    }

    override fun lastIndexOf(element: NBTBase<*>): Int {
        return value0.lastIndexOf(element)
    }

    override fun add(element: NBTBase<*>): Boolean {
        check(element)
        return value0.add(element)
    }

    override fun add(index: Int, element: NBTBase<*>) {
        check(element)
        return value0.add(index, element)
    }

    override fun addAll(index: Int, elements: Collection<NBTBase<*>>): Boolean {
        for (el in elements)
            check(el)
        return value0.addAll(index, elements)
    }

    override fun addAll(elements: Collection<NBTBase<*>>): Boolean {
        for (el in elements)
            check(el)
        return value0.addAll(elements)
    }

    override fun clear() {
        value0.clear()
    }

    override fun listIterator(): MutableListIterator<NBTBase<*>> {
        return value0.listIterator()
    }

    override fun listIterator(index: Int): MutableListIterator<NBTBase<*>> {
        return value0.listIterator(index)
    }

    override fun remove(element: NBTBase<*>): Boolean {
        check(element)
        return value0.remove(element)
    }

    override fun removeAll(elements: Collection<NBTBase<*>>): Boolean {
        for (el in elements)
            check(el)
        return value0.removeAll(elements)
    }

    override fun removeAt(index: Int): NBTBase<*> {
        return value0.removeAt(index)
    }

    override fun retainAll(elements: Collection<NBTBase<*>>): Boolean {
        for (el in elements)
            check(el)
        return value0.retainAll(elements)
    }

    override fun set(index: Int, element: NBTBase<*>): NBTBase<*> {
        check(element)
        return value0.set(index, element)
    }

    override fun subList(fromIndex: Int, toIndex: Int): MutableList<NBTBase<*>> {
        return value0.subList(fromIndex, toIndex)
    }
}
